File: frontend/src/lib/api/notifications.ts
// 站内通知（网关 /api/notifications*，服务端实现见后端 notifications_http.go）。
//
// 失败一律抛出（fetchApi 保留 HTTP 状态码）：通知列表界面上有独立的"加载失败"分支，
// 把"取不到"讲成"没有通知"是本项目刚修过的一类缺陷（口径见 lib/api/fields.ts 头部注释
// 与 docs-local/dev-log.md 的 v0.2.4 段），所以这里对 items 不做任何空数组兜底。
import { fetchApi } from "./client";
import { requireArray, safeCount } from "./fields";

export type NotificationType =
  | "comment.replied"
  | "entity.included"
  | "entity.review_approved"
  | "entity.review_rejected"
  | "import.completed";

export type NotificationSubjectType = "topic" | "entity" | "import" | "";

/**
 * 一行通知。payload 是各类型自己的键值袋（键随类型变化，且可能缺键），
 * 取值方一律按 possibly-undefined 处理，不要假定某个键一定在。
 */
export interface Notification {
  id: string;
  type: NotificationType;
  /** 系统事件（无操作者）时为空串：不要当用户名渲染，也不要造一个假名字。 */
  actor_id: string;
  actor_name: string;
  subject_type: NotificationSubjectType;
  subject_id: string;
  payload: Record<string, unknown>;
  /** >=1：同一聚合键的多条事件被服务端合并成一行，count 是合并条数。 */
  count: number;
  read: boolean;
  /** RFC3339。 */
  created_at: string;
  updated_at: string;
}

export interface NotificationListResponse {
  items: Notification[];
  total: number;
  unread: number;
}

export interface MarkNotificationReadResult {
  ok: boolean;
  unread: number;
}

export interface MarkAllNotificationsReadResult {
  ok: boolean;
  updated: number;
  unread: number;
}

/** 契约给的缺省与上限：limit 缺省 20、上限 100（超出会被服务端拒绝或截断，这里先夹住）。 */
export const NOTIFICATION_PAGE_SIZE_DEFAULT = 20;
export const NOTIFICATION_PAGE_SIZE_MAX = 100;

/**
 * 已读状态变化事件名：通知页改完已读后广播一次，顶栏角标据此**立刻**刷新，
 * 这样角标就不需要靠高频轮询去追（顶栏的定时刷新最短间隔 >=60s）。
 */
export const NOTIFICATIONS_CHANGED_EVENT = "mf:notifications-changed";

export function emitNotificationsChanged(): void {
  if (typeof window === "undefined") return;
  window.dispatchEvent(new Event(NOTIFICATIONS_CHANGED_EVENT));
}

function clampInt(value: unknown, fallback: number, min: number, max: number): number {
  const n = typeof value === "number" && Number.isFinite(value) ? Math.floor(value) : fallback;
  if (n < min) return min;
  if (n > max) return max;
  return n;
}

/** count 缺失或换型时按 1 处理：宁可不显示聚合条数，也不把坏值当条数展示。 */
function normalizeNotification(raw: Record<string, any>): Notification {
  const payload = raw?.payload;
  return {
    id: String(raw?.id ?? ""),
    type: (typeof raw?.type === "string" ? raw.type : "") as NotificationType,
    actor_id: typeof raw?.actor_id === "string" ? raw.actor_id : "",
    actor_name: typeof raw?.actor_name === "string" ? raw.actor_name : "",
    subject_type: (typeof raw?.subject_type === "string" ? raw.subject_type : "") as NotificationSubjectType,
    subject_id: typeof raw?.subject_id === "string" ? raw.subject_id : "",
    // payload 只是行内的次要信息（没有"空"文案），非对象时退回空袋；取值方本来就按缺键处理。
    payload: payload && typeof payload === "object" && !Array.isArray(payload) ? (payload as Record<string, unknown>) : {},
    count: safeCount(raw?.count, 1),
    read: raw?.read === true,
    created_at: typeof raw?.created_at === "string" ? raw.created_at : "",
    updated_at: typeof raw?.updated_at === "string" ? raw.updated_at : "",
  };
}

/** GET /notifications?limit=&offset= —— 按时间倒序，未登录 401 authentication_required。 */
export async function fetchNotifications(
  params: { limit?: number; offset?: number } = {}
): Promise<NotificationListResponse> {
  const limit = clampInt(params.limit, NOTIFICATION_PAGE_SIZE_DEFAULT, 1, NOTIFICATION_PAGE_SIZE_MAX);
  const offset = clampInt(params.offset, 0, 0, Number.MAX_SAFE_INTEGER);
  const res = await fetchApi<Record<string, any>>(`/notifications?limit=${limit}&offset=${offset}`);
  // 非数组即抛 invalid_response: items：界面上有"暂无通知"文案，漂移不能伪装成空列表。
  const items = requireArray<Record<string, any>>(res?.items, "items").map(normalizeNotification);
  return {
    items,
    total: safeCount(res?.total, items.length),
    unread: safeCount(res?.unread, 0),
  };
}

/**
 * GET /notifications/unread-count —— 只回未读总数的轻量端点（供顶栏角标）。
 * 直接回数字方便调用方；请求失败会抛出，由调用方决定要不要静默。
 */
export async function fetchUnreadCount(): Promise<number> {
  const res = await fetchApi<Record<string, any>>("/notifications/unread-count");
  return safeCount(res?.unread, 0);
}

/**
 * POST /notifications/:id/read —— 不是自己的或不存在时抛 404 not_found（ApiError.status === 404）。
 * 返回服务端算好的未读总数，调用方不必再猜。
 */
export async function markNotificationRead(id: string): Promise<MarkNotificationReadResult> {
  const res = await fetchApi<Record<string, any>>(`/notifications/${encodeURIComponent(id)}/read`, { method: "POST" });
  return { ok: res?.ok === true, unread: safeCount(res?.unread, 0) };
}

/** POST /notifications/read-all —— 返回本次更新行数与未读总数（应为 0）。 */
export async function markAllNotificationsRead(): Promise<MarkAllNotificationsReadResult> {
  const res = await fetchApi<Record<string, any>>("/notifications/read-all", { method: "POST" });
  return {
    ok: res?.ok === true,
    updated: safeCount(res?.updated, 0),
    unread: safeCount(res?.unread, 0),
  };
}

File: backend/internal/catalog/notifications.go
package catalog

// 站内通知（收件箱）。表结构与设计取舍见 backend/migrations/000003_notifications.up.sql 的注释，
// 完整报告见 docs-local/report-f1-notifications/REPORT.md。
//
// 落点：目录服务 —— 五类事件里四个产生端在目录（审核结果、收录、导入完成），
// 前端 /api 的主系统也是目录，而 catalog.user_preferences 已确立"目录库存按人裸 UUID 数据"的先例。
// 跨服务写入只剩一条："评论被回复"由互动服务按 POST /api/notifications/internal 投递
// （共享密钥 + 终端用户令牌），出站走互动服务自己的 internal/upstream。
//
// 三条不可动摇的语义：
//   1. 聚合：同一 (收件人, dedupe_key) 只有一行，count 累加、read_at 归零、updated_at 刷新。
//      "同一帖多条回复"因此是一行 + count，而不是一次刷屏。
//   2. 已读：read_at IS NULL = 未读；标记已读的 SQL 里带 recipient_id 条件。
//   3. 越权：列表/未读/标记已读一律以令牌身份为收件人，**不接受**调用方传 recipient_id；
//      对别人的通知标记已读与对不存在的 id 一样回 404 not_found（不泄露存在性）。

import (
	"context"
	"database/sql"
	"encoding/json"
	"fmt"
	"strings"
	"time"

	"github.com/google/uuid"
)

// 通知类型枚举（对外稳定码，与前端四语字典的键一一对应；只增不改）。
const (
	NotificationCommentReplied  = "comment.replied"
	NotificationEntityIncluded  = "entity.included"
	NotificationReviewApproved  = "entity.review_approved"
	NotificationReviewRejected  = "entity.review_rejected"
	NotificationImportCompleted = "import.completed"
)

// 列表分页口径：缺省 20、上限 100（越界静默收敛，与 /catalog/entities 风格一致）。
const (
	defaultNotificationLimit = 20
	maxNotificationLimit     = 100
)

// notificationTypes 的有序清单也是对外契约的一部分（NotificationTypes 用它）。
var (
	notificationTypeOrder = []string{
		NotificationCommentReplied, NotificationEntityIncluded, NotificationReviewApproved,
		NotificationReviewRejected, NotificationImportCompleted,
	}
	notificationTypes = map[string]bool{
		NotificationCommentReplied:  true,
		NotificationEntityIncluded:  true,
		NotificationReviewApproved:  true,
		NotificationReviewRejected:  true,
		NotificationImportCompleted: true,
	}
)

// NotificationTypes 返回全部可用类型码（供 openapi 描述与测试断言）。
func NotificationTypes() []string { return append([]string{}, notificationTypeOrder...) }

func validNotificationType(t string) bool { return notificationTypes[t] }

// Notification 是对外形状。dedupe_key 与 recipient_id 刻意不在其中：
// 前者是服务端的合并口径（不是展示字段），后者就是调用者本人。
type Notification struct {
	ID          string         `json:"id"`
	Type        string         `json:"type"`
	ActorID     string         `json:"actor_id"`
	ActorName   string         `json:"actor_name"`
	SubjectType string         `json:"subject_type"`
	SubjectID   string         `json:"subject_id"`
	Payload     map[string]any `json:"payload"`
	// Count 是这一行合并了多少条同键事件（>=1）。
	Count int  `json:"count"`
	Read  bool `json:"read"`
	// CreatedAt 是首条事件时间；UpdatedAt 是最近一次活动时间，列表按它倒序。
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

// NotificationInput 是服务端产生端填的内容。RecipientID 为空即丢弃（调用方已判定"没人要收"）。
type NotificationInput struct {
	RecipientID string
	Type        string
	ActorID     string
	ActorName   string
	SubjectType string
	SubjectID   string
	Payload     map[string]any
	// DedupeKey 为空时按 "type:subject_id" 生成：同类型的同一条落点合并成一行。
	DedupeKey string
	// EventID 是这条事件的稳定身份（跨服务投递传被回复的回复 id 之类）。
	// 同一 (收件人, DedupeKey) 再收到同一个 EventID 时**不再累加计数、不刷新未读**：
	// 上游重试（超时后其实已写入）不会把一条回复记成两条。为空则每次生成新 id（本服务内的事件）。
	EventID string
}

func (in NotificationInput) dedupe() string {
	if k := strings.TrimSpace(in.DedupeKey); k != "" {
		return k
	}
	return in.Type + ":" + in.SubjectID
}

// notificationWriter 让同一段 upsert 既能进调用方的事务（实体/关系写），也能直接走 DB（内部投递端点）。
type notificationWriter interface {
	ExecContext(ctx context.Context, query string, args ...any) (sql.Result, error)
}

// notificationUpsert 是**唯一**的写入语句：新增与合并走同一条，避免两套口径漂移。
// 合并时 read_at 归零（有新活动就要重新被看见）、count 累加、其余字段取最新事件。
const notificationUpsert = `
INSERT INTO catalog.notifications
  (id,recipient_id,type,actor_id,actor_name,subject_type,subject_id,payload,dedupe_key,last_event_id,count,created_at,updated_at)
VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,1,now(),now())
ON CONFLICT (recipient_id,dedupe_key) DO UPDATE SET
  type = EXCLUDED.type,
  actor_id = EXCLUDED.actor_id,
  actor_name = EXCLUDED.actor_name,
  subject_type = EXCLUDED.subject_type,
  subject_id = EXCLUDED.subject_id,
  payload = EXCLUDED.payload,
  -- 重试幂等：同一事件再投一次时计数与未读都不动（否则一次回复会被记成两条）。
  count = CASE WHEN catalog.notifications.last_event_id IS DISTINCT FROM EXCLUDED.last_event_id
               THEN catalog.notifications.count + 1 ELSE catalog.notifications.count END,
  read_at = CASE WHEN catalog.notifications.last_event_id IS DISTINCT FROM EXCLUDED.last_event_id
                 THEN NULL ELSE catalog.notifications.read_at END,
  updated_at = CASE WHEN catalog.notifications.last_event_id IS DISTINCT FROM EXCLUDED.last_event_id
                    THEN now() ELSE catalog.notifications.updated_at END,
  last_event_id = EXCLUDED.last_event_id`

// notify 写/合并一条通知。类型是代码常量，写错了要失败得响一点（进测试眼），
// 因此非法类型返回错误而不是静默丢弃。
func notify(ctx context.Context, q notificationWriter, in NotificationInput) error {
	if strings.TrimSpace(in.RecipientID) == "" {
		return nil
	}
	if !validNotificationType(in.Type) {
		return fmt.Errorf("invalid_notification_type: %s", in.Type)
	}
	payload := in.Payload
	if payload == nil {
		payload = map[string]any{}
	}
	event := strings.TrimSpace(in.EventID)
	if event == "" {
		event = uuid.NewString()
	}
	_, err := q.ExecContext(ctx, notificationUpsert,
		uuid.NewString(), in.RecipientID, in.Type, nullable(in.ActorID), in.ActorName,
		in.SubjectType, in.SubjectID, encode(payload), in.dedupe(), event)
	return err
}

const notificationColumns = `id::text, type, COALESCE(actor_id::text,''), actor_name, subject_type, subject_id,
       payload, count, (read_at IS NOT NULL) AS read, created_at, updated_at`

// ListNotifications 是收件箱列表：按最近活动倒序，返回 (条目, 总数, 未读数)。
// 未读数与列表在同一次调用里返回是有意的：列表页两个数要一致，分两次请求会出现
// "角标 3、列表全已读"的中间窗口。
func (s *Store) ListNotifications(ctx context.Context, recipientID string, limit, offset int) ([]Notification, int, int, error) {
	if limit <= 0 || limit > maxNotificationLimit {
		limit = defaultNotificationLimit
	}
	if offset < 0 {
		offset = 0
	}
	rows, err := s.DB.QueryContext(ctx,
		"SELECT "+notificationColumns+" FROM catalog.notifications WHERE recipient_id=$1 ORDER BY updated_at DESC, id DESC LIMIT $2 OFFSET $3",
		recipientID, limit, offset)
	if err != nil {
		return nil, 0, 0, err
	}
	defer rows.Close()
	items := []Notification{}
	for rows.Next() {
		var n Notification
		var read bool
		var raw []byte
		if err := rows.Scan(&n.ID, &n.Type, &n.ActorID, &n.ActorName, &n.SubjectType, &n.SubjectID,
			&raw, &n.Count, &read, &n.CreatedAt, &n.UpdatedAt); err != nil {
			return nil, 0, 0, err
		}
		n.Read = read
		if len(raw) > 0 {
			if err := json.Unmarshal(raw, &n.Payload); err != nil {
				// payload 是本服务写进去的 JSON 对象，读不回来是服务端故障，
				// 不能折成空 payload 让调用方以为"这条通知没内容"。
				return nil, 0, 0, fmt.Errorf("invalid_notification_payload: %w", err)
			}
		}
		if n.Payload == nil {
			n.Payload = map[string]any{}
		}
		items = append(items, n)
	}
	if err := rows.Err(); err != nil {
		return nil, 0, 0, err
	}
	var total int
	if err := s.DB.QueryRowContext(ctx, "SELECT count(*) FROM catalog.notifications WHERE recipient_id=$1", recipientID).Scan(&total); err != nil {
		return nil, 0, 0, err
	}
	unread, err := s.NotificationUnreadCount(ctx, recipientID)
	if err != nil {
		return nil, 0, 0, err
	}
	return items, total, unread, nil
}

// NotificationUnreadCount 是角标端点唯一要的一次聚合：走 notifications_unread 部分索引，
// 不取行、不 JOIN、不解析 payload。
func (s *Store) NotificationUnreadCount(ctx context.Context, recipientID string) (int, error) {
	var n int
	err := s.DB.QueryRowContext(ctx,
		"SELECT count(*) FROM catalog.notifications WHERE recipient_id=$1 AND read_at IS NULL", recipientID).Scan(&n)
	return n, err
}

// MarkNotificationRead 标记单条已读（已读时间戳保留首次值），返回标记后的未读数。
// 收件人条件进 WHERE：别人的通知与不存在的 id 都 0 行 → sql.ErrNoRows，HTTP 层统一回 404 not_found。
func (s *Store) MarkNotificationRead(ctx context.Context, recipientID, id string) (int, error) {
	if _, err := uuid.Parse(id); err != nil {
		return 0, sql.ErrNoRows
	}
	res, err := s.DB.ExecContext(ctx,
		"UPDATE catalog.notifications SET read_at=COALESCE(read_at, now()) WHERE id=$1 AND recipient_id=$2", id, recipientID)
	if err != nil {
		return 0, err
	}
	if n, _ := res.RowsAffected(); n == 0 {
		return 0, sql.ErrNoRows
	}
	return s.NotificationUnreadCount(ctx, recipientID)
}

// MarkAllNotificationsRead 标记全部未读，返回本次标记的条数与之后的未读数
// （后者仍回读一次：返回的 0 应当是库里的 0，不是"我假设它是 0"）。
func (s *Store) MarkAllNotificationsRead(ctx context.Context, recipientID string) (int64, int, error) {
	res, err := s.DB.ExecContext(ctx,
		"UPDATE catalog.notifications SET read_at=now() WHERE recipient_id=$1 AND read_at IS NULL", recipientID)
	if err != nil {
		return 0, 0, err
	}
	updated, _ := res.RowsAffected()
	unread, err := s.NotificationUnreadCount(ctx, recipientID)
	return updated, unread, err
}
